rpyc/rpyc_client/command_handlers.py

```diff
@@ -0,0 +1,203 @@
+import psutil
+import datetime
+
+"""
+All the command handlers.
+They parse, execute and possibly format the output for the terminal.
+They (must) also return (and receive) additional information to the command_handlers_manager. 
+"""
+
+
+class CommandHandler(object):
+    def __init__(self, rpyc_conn):
+        self.rpyc_conn = rpyc_conn
+        self.empty_output = (None, None)
+
+    def _parse_input(self, split_input, additional_input):
+        pass
+
+    def _execute(self, **kwargs):
+        return self.empty_output
+
+    def _format_output(self, output):
+        return output
+
+    def handle_command(self, split_input, additional_input):
+        try:
+            kwargs = self._parse_input(split_input, additional_input)
+            terminal_output, manager_output = self._execute(**kwargs)
+            terminal_output = self._format_output(terminal_output) if terminal_output else terminal_output
+            return terminal_output, manager_output
+        except BaseException as e:
+            return 'Error: ' + str(e), None
+
+
+class CopyFileHandler(CommandHandler):
+    def _parse_input(self, split_input, additional_input):
+        if len(split_input) != 3:
+            raise TypeError('expected two arguments')
+
+        command = split_input[0]
+        if command == 'upload':
+            return {'local_path': split_input[1], 'remote_path': split_input[2], 'mod': command}
+        else:
+            return {'remote_path': split_input[1], 'local_path': split_input[2], 'mod': command}
+
+    def _execute(self, local_path, remote_path, mod):
+        chunk_size = 16000
+        local_path_open_mod, remote_path_open_mod = ("rb", "wb") if mod == 'upload' else ("wb", "rb")
+
+        with open(local_path, local_path_open_mod) as lf:
+            with self.rpyc_conn.builtin.open(remote_path, remote_path_open_mod) as rf:
+                if mod == 'download':  # remote to local
+                    self.__copy_file(rf, lf, chunk_size)
+                elif mod == 'upload':  # local to remote
+                    self.__copy_file(lf, rf, chunk_size)
+
+        return self.empty_output
+
+    def __copy_file(self, src_file_handler, dst_file_handler, chunk_size):
+        while True:
+            buf = src_file_handler.read(chunk_size)
+            if not buf:
+                break
+            dst_file_handler.write(buf)
+
+
+class DirlistHandler(CommandHandler):
+    def _parse_input(self, split_input, additional_input):
+        if len(split_input) != 2:
```

Comment:
this is really code duplication, I believe you can export it to a function and call it with the right parameters every time

Reply:
Tried to do this as much as possible